test/e2e/utils_test.go

```diff
@@ -167,3 +191,163 @@ func verifyDeploymentArgs(k8sclient *kubernetes.Clientset, deploymentName string
+	_, err = subscriptionClient.Patch(ctx, subName, types.MergePatchType, payload, v1.PatchOptions{})
+	return err
+}
+
+// waitForCertificateReadiness polls the status of the Certificate object and returns non-nil error
+// once the Ready condition is true, otherwise should return a time-out error
+func waitForCertificateReadiness(ctx context.Context, certName, namespace string) error {
+	return wait.PollImmediate(PollInterval, TestTimeout, func() (bool, error) {
+		cert, err := certmanagerClient.CertmanagerV1().Certificates(namespace).Get(ctx, certName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		for _, cond := range cert.Status.Conditions {
+			if cond.Type == cmv1.CertificateConditionReady {
+				return cond.Status == cmmetav1.ConditionTrue, nil
+			}
+		}
+		return false, nil
+	})
+}
+
+// verifyCertificate loads the tls secret as a X509 certificate and verifies the following
+// - certificate secret is non null, i.e. secret contains "tls.crt", "tls.key" keys
+// - certificate hasn't expired
+// - certificate has subject CN matching provided hostname
+func verifyCertificate(ctx context.Context, secretName, namespace, hostname string) (bool, error) {
+	var isVerified bool
+	err := wait.PollImmediate(PollInterval, TestTimeout, func() (bool, error) {
+		secret, err := loader.KubeClient.CoreV1().Secrets(namespace).Get(ctx, secretName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		isVerified, err = library.VerifyCertificate(secret, hostname)
+		if err != nil {
+			return true, err
+		}
+		return true, nil
+	})
+	return isVerified, err
+}
+
+// verifyCertificateRenewed repeatedly loads the tls secret as a X509 certificate every pollDuration
+// and returns true if certificate was renewed at least once
+func verifyCertificateRenewed(ctx context.Context, secretName, namespace string, pollDuration time.Duration) (bool, error) {
+	var wasRenewedAtleastOnce bool
+	var initExpiryTime *time.Time
+	err := wait.PollImmediate(pollDuration, TestTimeout, func() (bool, error) {
+		secret, err := loader.KubeClient.CoreV1().Secrets(namespace).Get(ctx, secretName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		cert, err := library.GetX509Certificate(secret)
+		if err != nil {
+			return false, err
+		}
+
+		// cert expiry time is recorded upon initial run
+		if initExpiryTime == nil {
+			initExpiryTime = &cert.NotAfter
+		}
+
+		// checks if expiry time was updated
+		if *initExpiryTime == cert.NotAfter {
+			return false, nil
+		}
+
+		// iff expiry time was updated, check if new expiry is not ahead
+		// return an error, else certificate was renewed properly
+		if !cert.NotAfter.After(*initExpiryTime) {
+			return false, fmt.Errorf("previous expiry time of the certificate cannot be ahead of the current expiry time")
+		}
+
+		wasRenewedAtleastOnce = true
+		return true, nil
+	})
+	return wasRenewedAtleastOnce, err
+}
```

Comment:
Same here as that of `verifyCertificate`.

Reply:
Done.